File: application/app/src/main/java/com/krs/community/app/AppController.java
package com.krs.community.app;


import android.app.Application;
import android.content.Context;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.graphics.Typeface;
import android.os.StrictMode;
import android.text.TextUtils;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.content.res.ResourcesCompat;
import androidx.multidex.MultiDex;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.ImageLoader;
import com.android.volley.toolbox.Volley;
import com.crashlytics.android.Crashlytics;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.krs.community.R;
import com.krs.community.utils.AppConstants;
import com.krs.community.utils.ConnectivityReceiver;
import com.krs.community.utils.LocaleHelper;
import com.krs.community.volley.LruBitmapCache;

import java.util.ArrayList;

import io.fabric.sdk.android.Fabric;

public class AppController extends Application {

    private static final String TAG = AppController.class.getSimpleName();

    private static AppController mInstance;
    public boolean isUpdate = false;
    //public FirebaseAnalytics firebaseAnalytics;
    /*public ArrayAdapter<String> dataAdapter;
    public List<String> lstGotra;*/
    public ArrayList<String> lstGotra;
    public ArrayList<String> lstNative;
    public ArrayList<String> lstEducation;

    public SharedPreferences mSharedPreferences;
    public SharedPreferences.Editor mEditor;
    ConnectivityReceiver broadcastRevcevier;
    private RequestQueue mRequestQueue;
    public GoogleSignInClient mGoogleSignInClient;
    private ImageLoader mImageLoader;
    LruBitmapCache mLruBitmapCache;
    public Typeface typeface,typeface_bold;

    public static synchronized AppController getInstance() {
        return mInstance;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        typeface = ResourcesCompat.getFont(getApplicationContext(),R.font.montserrat_regular);
        typeface_bold = ResourcesCompat.getFont(getApplicationContext(),R.font.montserrat_semibold);

        final Fabric fabric = new Fabric.Builder(this).kits(new Crashlytics()).debuggable(true).build();
        Fabric.with(fabric);

        MultiDex.install(this);
        mInstance = this;
        lstGotra =new ArrayList<>();
        lstNative =new ArrayList<>();
        lstEducation =new ArrayList<>();

        mSharedPreferences = getSharedPreferences(AppConstants.PREF_NAME, MODE_PRIVATE);
        mEditor = mSharedPreferences.edit();
        initFirebaseAnalytics();
        StrictMode.VmPolicy.Builder builder = new StrictMode.VmPolicy.Builder();
        StrictMode.setVmPolicy(builder.build());
        broadcastRevcevier = new ConnectivityReceiver();
        registerReceiver(broadcastRevcevier, new IntentFilter("android.net.conn.CONNECTIVITY_CHANGE"));

        Log.d(TAG,"AppController Screen");
        mEditor.putBoolean(getString(R.string.app_create),true);
        mEditor.apply();

        GoogleSignInOptions gso = new GoogleSignInOptions
                .Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.web_client_id))
                .requestEmail().build();
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);


    }

    public ImageLoader getImageLoader() {
        getRequestQueue();
        if (mImageLoader == null) {
            getLruBitmapCache();
            mImageLoader = new ImageLoader(this.mRequestQueue, mLruBitmapCache);
        }

        return this.mImageLoader;
    }

    public LruBitmapCache getLruBitmapCache() {
        if (mLruBitmapCache == null)
            mLruBitmapCache = new LruBitmapCache();
        return this.mLruBitmapCache;
    }



    @Override
    public void onTerminate() {
        super.onTerminate();
        if (broadcastRevcevier != null) {
            unregisterReceiver(broadcastRevcevier);
            broadcastRevcevier = null;
        }
    }

    public void setConnectivityListener(ConnectivityReceiver.ConnectivityReceiverListener listener) {
        ConnectivityReceiver.connectivityReceiverListener = listener;
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(LocaleHelper.onAttach(base, "en"));
    }

    private void initFirebaseAnalytics() {
       /* firebaseAnalytics = FirebaseAnalytics.getInstance(this);

        //Set whether analytics collection is enabled for this app on this device.
        firebaseAnalytics.setAnalyticsCollectionEnabled(true);

        //Set the minimum engagement time required before starting a session.
        firebaseAnalytics.setMinimumSessionDuration(2000);

        //Set the duration of inactivity that terminates the current session. The default value is 1800000 (30 minutes).
        firebaseAnalytics.setSessionTimeoutDuration(300000);*/
    }

       public RequestQueue getRequestQueue() {
        if (mRequestQueue == null) {
            mRequestQueue = Volley.newRequestQueue(getApplicationContext());
        }

        return mRequestQueue;
    }

    public <T> void addToRequestQueue(@NonNull Request<T> req, String tag) {
        // set the default tag if tag is empty
        req.setTag(TextUtils.isEmpty(tag) ? TAG : tag);
        getRequestQueue().add(req);
    }

    public <T> void addToRequestQueue(@NonNull Request<T> req) {
        req.setTag(TAG);
        getRequestQueue().add(req);
    }




}

File: application/app/src/main/java/com/krs/community/app/AppController.kt
package com.krs.community.app


import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.graphics.Typeface
import android.os.Build
import android.os.Handler
import android.os.StrictMode
import android.util.Log
import androidx.core.content.res.ResourcesCompat
import androidx.multidex.BuildConfig
import androidx.multidex.MultiDex
import com.crashlytics.android.Crashlytics
import com.downloader.PRDownloader
import com.downloader.PRDownloaderConfig
import com.facebook.FacebookSdk
import com.facebook.appevents.AppEventsLogger
import com.facebook.drawee.backends.pipeline.Fresco
import com.github.squti.guru.Guru
import com.github.squti.guru.GuruConfig
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.krs.community.R
import com.krs.community.app.ConnectionLiveData.Companion.isNetworkConnected
import com.krs.community.repositories.*
import com.krs.community.retrofit.ApiServices
import com.krs.community.retrofit.RetrofitBase
import com.krs.community.utils.AppConstants
import com.krs.community.utils.Coroutines
import com.krs.community.utils.LocaleHelper
import com.krs.community.viewmodelfactory.*
import io.fabric.sdk.android.Fabric
import net.gotev.uploadservice.UploadServiceConfig
import org.json.JSONObject
import org.kodein.di.Kodein
import org.kodein.di.KodeinAware
import org.kodein.di.android.x.androidXModule
import org.kodein.di.generic.bind
import org.kodein.di.generic.instance
import org.kodein.di.generic.provider
import org.kodein.di.generic.singleton


class AppController : Application(), KodeinAware {

    lateinit var mGoogleSignInClient: GoogleSignInClient
    lateinit var typeface: Typeface
    lateinit var firebaseAnalytics: FirebaseAnalytics
    lateinit var logger: AppEventsLogger
    lateinit var typeface_bold: Typeface
    lateinit var retrofitBase: RetrofitBase
    lateinit var connectionLiveData: ConnectionLiveData
    var start: Int = 0
    val length: Int = 50
    val mHandler: Handler = Handler()


    companion object {
        val TAG = AppController::class.java.simpleName
        lateinit var mApplication: AppController
        const val notificationChannelID = "TestChannel"
        val INTERVAL = 1000 * 60 * 3 //3 minutes
    }

    override val kodein = Kodein.lazy {

        import(androidXModule(this@AppController))

        bind() from singleton { ApiServices() }
        bind() from singleton { AppDatabase(instance()) }

        bind() from singleton { RegisterRepository(instance()) }
        bind() from singleton { LoginRepository(instance()) }
        bind() from singleton { PasswordRepository(instance()) }
        bind() from singleton { ShareEventRepository(instance()) }
        bind() from singleton { BrowseCityRepository(instance(), instance()) }
        bind() from singleton { ByDistanceRepository(instance(), instance()) }
        bind() from singleton { FamilyDetailRepository(instance()) }
        bind() from singleton { ProfileDetailRepository(instance(), instance()) }
        bind() from singleton { DashboardRepository(instance(), instance()) }
        bind() from singleton { StatisticsRepository(instance(), instance()) }
        bind() from singleton { ContactListRepository(instance(), instance()) }
        bind() from singleton { SmartSearchRepository(instance(), instance()) }
        bind() from singleton { SmartFilterRepository(instance(), instance()) }
        bind() from singleton { DocumentListRepository(instance()) }
        bind() from singleton { CalendarSearchRepository(instance(), instance()) }
        bind() from singleton { NewsRepository(instance()) }
        bind() from singleton { RoomMemberRepository(instance(), instance()) }
        bind() from singleton { CommitteeRepository(instance(), instance()) }

        bind() from provider { ContactListViewModelFactory(instance()) }
        bind() from provider { StatisticsViewModelFactory(instance()) }
        bind() from provider { FamilyDetailViewModelFactory(instance()) }
        bind() from provider { RegisterViewModelFactory(instance()) }
        bind() from provider { LoginViewModelFactory(instance()) }
        bind() from provider { PasswordViewModelFactory(instance()) }
        bind() from provider { ShareEventViewModelFactory(instance()) }
        bind() from provider { BrowseCityViewModelFactory(instance()) }
        bind() from provider { ByDistanceViewModelFactory(instance()) }
        bind() from provider { ProfileDetailViewModelFactory(instance()) }
        bind() from provider { DashboardViewModelFactory(instance()) }
        bind() from provider { SmartSearchViewModelFactory(instance()) }
        bind() from provider { DocumentListViewModelFactory(instance()) }
        bind() from provider { SmartFilterViewModelFactory(instance()) }
        bind() from provider { CalendarSearchViewModelFactory(instance()) }
        bind() from provider { NewsModelFactory(instance()) }
        bind() from provider { RoomMemberViewModelFactory(instance()) }
        bind() from provider { CommiteeViewModelFactory(instance()) }

    }

    fun firebaseAnalytics(getContext: Context?, Name: String?) {
        firebaseAnalytics = FirebaseAnalytics.getInstance(getContext!!)
        firebaseAnalytics.setCurrentScreen((getContext as Activity?)!!, "Screen", Name)
    }

    fun facebookAnalytics(getContext: Context?, Name: String?) {
        logger = AppEventsLogger.newLogger(getContext)
        logger.logEvent(Name)
    }

    /*  fun stringTranslateAPI(Name: String?):String {
          translateAPI = TranslateAPI(Language.AUTO_DETECT, Language.TAMIL, Name);

          translateAPI.setTranslateListener(object :TranslateAPI.TranslateListener{
              override fun onSuccess(translatedText: String?) : String? {
                  Log.d(TAG, "onSuccess: " + translatedText);
                  return translatedText
              }

              override fun onFailure(ErrorText: String?): String? {
                  Log.d(TAG, "onSuccess: " + ErrorText);
                  return ErrorText
              }

          })
          return ""
      }*/

    private val mHandlerTask = object : Runnable {
        override fun run() {
            if (Guru.getBoolean(getString(R.string.policy), false)) {
                if (isNetworkConnected(this@AppController)) {
                    updateUserStatus()
                }
                mHandler.postDelayed(this, INTERVAL.toLong())
            }
        }
    }

    private fun startRepeatingTask() {
        mHandlerTask.run()
    }

    private fun stopRepeatingTask() {
        mHandler.removeCallbacks(mHandlerTask)
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= 26) {
            val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val channel = NotificationChannel(notificationChannelID, "TestApp Channel", NotificationManager.IMPORTANCE_LOW)
            manager.createNotificationChannel(channel)
        }
    }

    @SuppressLint("CommitPrefEdits")
    override fun onCreate() {
        super.onCreate()

        mApplication = this
        connectionLiveData = ConnectionLiveData(this)
        FacebookSdk.sdkInitialize(applicationContext)
        createNotificationChannel()

        UploadServiceConfig.initialize(
                context = this,
                defaultNotificationChannel = notificationChannelID,
                debug = BuildConfig.DEBUG
        )

        typeface = ResourcesCompat.getFont(applicationContext, R.font.montserrat_regular)!!
        typeface_bold = ResourcesCompat.getFont(applicationContext, R.font.montserrat_semibold)!!

        retrofitBase = RetrofitBase(this, false)
        Fresco.initialize(applicationContext)

        Fabric.with(this, Crashlytics())

        /* val fabric = Fabric.Builder(this).kits(Crashlytics()).debuggable(true).build()
         Fabric.with(fabric)*/

        MultiDex.install(this)

        GuruConfig.initDefault(GuruConfig.Builder()
                .setFileName(AppConstants.PREF_NAME)
                .setMode(Context.MODE_PRIVATE)
                .build())


        val builder = StrictMode.VmPolicy.Builder()
        StrictMode.setVmPolicy(builder.build())


        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.web_client_id))
                .requestEmail().build()
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso)

        val config = PRDownloaderConfig.newBuilder()
                .setDatabaseEnabled(true)
                .setReadTimeout(30_000)
                .setConnectTimeout(30_000)
                .build()
        PRDownloader.initialize(applicationContext, config)

        startRepeatingTask()
        Log.v("AppControler", "onCreate")
    }


    override fun onTerminate() {
        super.onTerminate()
        Log.v("AppControler", "onTerminate")
        stopRepeatingTask()
    }


    private fun updateUserStatus() {
        Coroutines.io {
            val memberId = Guru.getString(getString(R.string.member_id), "")
            if (!memberId.isNullOrEmpty()) {
                val jsonObject = JSONObject()
                jsonObject.put(getString(R.string.id), memberId)
                jsonObject.put(getString(R.string.user_id), Guru.getString(getString(R.string.user_id), ""))
                jsonObject.put(getString(R.string.access_token), Guru.getString(getString(R.string.access_token), ""))
                val updated = JsonParser().parse(jsonObject.toString()) as JsonObject
                retrofitBase.apiServices.getUserStatus(updated)
            }
        }
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(LocaleHelper.onAttach(base, "en"))
    }

}